filebeat/input/filestream/internal/task/group.go

```diff
@@ -0,0 +1,129 @@
+package task
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"golang.org/x/sync/semaphore"
+)
+
+type Group struct {
+	limit uint64
+
+	sem *semaphore.Weighted
+	wg  *sync.WaitGroup
+
+	errs     []error
+	errsSize uint64
+	errsNext uint64
+	errsMu   *sync.Mutex
+
+	ctx       context.Context
+	cancelCtx context.CancelFunc
+}
+
+type Goer interface {
+	Go(fn func(context.Context) error) error
+}
+
+func NewGroup(limit uint64, maxErrors uint64) *Group {
+	if maxErrors < 1 {
+		maxErrors = 1
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+
+	g := Group{
+		limit:     limit,
+		wg:        &sync.WaitGroup{},
+		errsSize:  maxErrors,
+		errs:      make([]error, maxErrors),
+		errsMu:    &sync.Mutex{},
+		ctx:       ctx,
+		cancelCtx: cancel,
+	}
+
+	if limit > 0 {
+		g.sem = semaphore.NewWeighted(int64(limit))
+	}
+
+	return &g
+}
+
+// Go starts fn on a goroutine when a worker becomes available.
+// If the worker pool was already closed, Go returns a context.Canceled error.
+// If there are no workers available and Group.Stop() is called, fn is discarded.
+// Go does not block.
+func (p *Group) Go(fn func(context.Context) error) error {
+	if err := p.ctx.Err(); err != nil {
+		return fmt.Errorf("task group is closed: %w", err)
+	}
+
+	p.wg.Add(1)
+	go func() {
+		defer p.wg.Done()
+
+		if p.limit != 0 {
+			defer p.sem.Release(1)
```

Comment:
Should this statement be after the error check for `p.sem.Acquire`?  In other words, if `p.sem.Acquire` returns an error, I don't think the internal weight in `p.sem` is incremented so decrementing it with `p.sem.Release` may not be a good idea?